`ConcurrentModificationException` in `DefaultObjectWriter.write`

Error logs:
```
[2024-08-07 19:45:59,295] ERROR Unexpected exception when commit stream set object (com.automq.stream.s3.S3Storage)
java.util.concurrent.CompletionException: java.util.ConcurrentModificationException
        at java.base/java.util.concurrent.CompletableFuture.encodeThrowable(CompletableFuture.java:315)
        at java.base/java.util.concurrent.CompletableFuture.completeThrowable(CompletableFuture.java:320)
        at java.base/java.util.concurrent.CompletableFuture$UniAccept.tryFire(CompletableFuture.java:722)
        at java.base/java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:510)
        at java.base/java.util.concurrent.CompletableFuture.complete(CompletableFuture.java:2147)
        at com.automq.stream.utils.AsyncRateLimiter.tick(AsyncRateLimiter.java:68)
        at com.automq.stream.utils.ThreadUtils.lambda$wrapRunnable$0(ThreadUtils.java:74)
        at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:539)
        at java.base/java.util.concurrent.FutureTask.runAndReset(FutureTask.java:305)
        at java.base/java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:305)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
        at java.base/java.lang.Thread.run(Thread.java:840)
Caused by: java.util.ConcurrentModificationException
        at java.base/java.util.LinkedList$ListItr.checkForComodification(LinkedList.java:970)
        at java.base/java.util.LinkedList$ListItr.next(LinkedList.java:892)
        at com.automq.stream.s3.ObjectWriter$DefaultObjectWriter.tryUploadPart(ObjectWriter.java:135)
        at com.automq.stream.s3.ObjectWriter$DefaultObjectWriter.write(ObjectWriter.java:108)
        at com.automq.stream.s3.DeltaWALUploadTask.lambda$upload0$4(DeltaWALUploadTask.java:153)
        at java.base/java.util.concurrent.CompletableFuture$UniAccept.tryFire(CompletableFuture.java:718)
        ... 10 more
[2024-08-07 19:45:59,295] ERROR upload delta WAL fail (com.automq.stream.s3.S3Storage)
java.util.concurrent.CompletionException: java.util.ConcurrentModificationException
        at java.base/java.util.concurrent.CompletableFuture.encodeThrowable(CompletableFuture.java:315)
        at java.base/java.util.concurrent.CompletableFuture.completeThrowable(CompletableFuture.java:320)
        at java.base/java.util.concurrent.CompletableFuture$UniAccept.tryFire(CompletableFuture.java:722)
        at java.base/java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:510)
        at java.base/java.util.concurrent.CompletableFuture.complete(CompletableFuture.java:2147)
        at com.automq.stream.utils.AsyncRateLimiter.tick(AsyncRateLimiter.java:68)
        at com.automq.stream.utils.ThreadUtils.lambda$wrapRunnable$0(ThreadUtils.java:74)
        at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:539)
        at java.base/java.util.concurrent.FutureTask.runAndReset(FutureTask.java:305)
        at java.base/java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:305)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
        at java.base/java.lang.Thread.run(Thread.java:840)
Caused by: java.util.ConcurrentModificationException
        at java.base/java.util.LinkedList$ListItr.checkForComodification(LinkedList.java:970)
        at java.base/java.util.LinkedList$ListItr.next(LinkedList.java:892)
        at com.automq.stream.s3.ObjectWriter$DefaultObjectWriter.tryUploadPart(ObjectWriter.java:135)
        at com.automq.stream.s3.ObjectWriter$DefaultObjectWriter.write(ObjectWriter.java:108)
        at com.automq.stream.s3.DeltaWALUploadTask.lambda$upload0$4(DeltaWALUploadTask.java:153)
        at java.base/java.util.concurrent.CompletableFuture$UniAccept.tryFire(CompletableFuture.java:718)
        ... 10 more
```

There're also memory leaks:
```
[2024-08-07 19:45:54,534] ERROR LEAK: ByteBuf.release() was not called before it's garbage-collected. See https://netty.io/wiki/reference-counted-objects.html for more information.
Recent access records:
Created at:
        io.netty.buffer.PooledByteBufAllocator.newDirectBuffer(PooledByteBufAllocator.java:404)
        io.netty.buffer.AbstractByteBufAllocator.directBuffer(AbstractByteBufAllocator.java:188)
        io.netty.buffer.AbstractByteBufAllocator.directBuffer(AbstractByteBufAllocator.java:179)
        com.automq.stream.s3.ByteBufAlloc.byteBuffer(ByteBufAlloc.java:134)
        com.automq.stream.s3.ObjectWriter$DataBlock.<init>(ObjectWriter.java:246)
        com.automq.stream.s3.ObjectWriter$DefaultObjectWriter.lambda$write$0(ObjectWriter.java:103)
        java.base/java.lang.Iterable.forEach(Iterable.java:75)
        com.automq.stream.s3.ObjectWriter$DefaultObjectWriter.write(ObjectWriter.java:102)
        com.automq.stream.s3.DeltaWALUploadTask.lambda$upload0$4(DeltaWALUploadTask.java:153)
        java.base/java.util.concurrent.CompletableFuture.uniAcceptNow(CompletableFuture.java:757)
        java.base/java.util.concurrent.CompletableFuture.uniAcceptStage(CompletableFuture.java:735)
        java.base/java.util.concurrent.CompletableFuture.thenAccept(CompletableFuture.java:2182)
        com.automq.stream.s3.DeltaWALUploadTask.upload0(DeltaWALUploadTask.java:153)
        com.automq.stream.s3.DeltaWALUploadTask.lambda$upload$1(DeltaWALUploadTask.java:122)
        com.automq.stream.utils.FutureUtil.exec(FutureUtil.java:77)
        com.automq.stream.s3.DeltaWALUploadTask.lambda$upload$2(DeltaWALUploadTask.java:122)
        java.base/java.util.concurrent.CompletableFuture$UniAccept.tryFire(CompletableFuture.java:718)
        java.base/java.util.concurrent.CompletableFuture$Completion.run(CompletableFuture.java:482)
        java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
        java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
        java.base/java.lang.Thread.run(Thread.java:840) (io.netty.util.ResourceLeakDetector)
[2024-08-07 19:45:54,534] ERROR LEAK: ByteBuf.release() was not called before it's garbage-collected. See https://netty.io/wiki/reference-counted-objects.html for more information.
Recent access records:
Created at:
        io.netty.buffer.AbstractByteBufAllocator.compositeDirectBuffer(AbstractByteBufAllocator.java:224)
        com.automq.stream.s3.ByteBufAlloc.compositeByteBuffer(ByteBufAlloc.java:110)
        com.automq.stream.s3.ObjectWriter$DataBlock.<init>(ObjectWriter.java:245)
        com.automq.stream.s3.ObjectWriter$DefaultObjectWriter.lambda$write$0(ObjectWriter.java:103)
        java.base/java.lang.Iterable.forEach(Iterable.java:75)
        com.automq.stream.s3.ObjectWriter$DefaultObjectWriter.write(ObjectWriter.java:102)
        com.automq.stream.s3.DeltaWALUploadTask.lambda$upload0$4(DeltaWALUploadTask.java:153)
        java.base/java.util.concurrent.CompletableFuture.uniAcceptNow(CompletableFuture.java:757)
        java.base/java.util.concurrent.CompletableFuture.uniAcceptStage(CompletableFuture.java:735)
        java.base/java.util.concurrent.CompletableFuture.thenAccept(CompletableFuture.java:2182)
        com.automq.stream.s3.DeltaWALUploadTask.upload0(DeltaWALUploadTask.java:153)
        com.automq.stream.s3.DeltaWALUploadTask.lambda$upload$1(DeltaWALUploadTask.java:122)
        com.automq.stream.utils.FutureUtil.exec(FutureUtil.java:77)
        com.automq.stream.s3.DeltaWALUploadTask.lambda$upload$2(DeltaWALUploadTask.java:122)
        java.base/java.util.concurrent.CompletableFuture$UniAccept.tryFire(CompletableFuture.java:718)
        java.base/java.util.concurrent.CompletableFuture$Completion.run(CompletableFuture.java:482)
        java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
        java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
        java.base/java.lang.Thread.run(Thread.java:840) (io.netty.util.ResourceLeakDetector)
```

--- a/s3stream/src/main/java/com/automq/stream/s3/ObjectWriter.java
+++ b/s3stream/src/main/java/com/automq/stream/s3/ObjectWriter.java
@@ -155,7 +155,7 @@ public interface ObjectWriter {
             }
         }
 
-        public CompletableFuture<Void> close() {
+        public synchronized CompletableFuture<Void> close() {
             CompositeByteBuf buf = ByteBufAlloc.compositeByteBuffer();
             for (DataBlock block : waitingUploadBlocks) {
                 buf.addComponent(true, block.buffer());

--- a/s3stream/src/main/java/com/automq/stream/s3/ObjectWriter.java
+++ b/s3stream/src/main/java/com/automq/stream/s3/ObjectWriter.java
@@ -99,11 +99,11 @@ public interface ObjectWriter {
 
         public synchronized void write(long streamId, List<StreamRecordBatch> records) {
             List<List<StreamRecordBatch>> blocks = groupByBlock(records);
-            blocks.forEach(blockRecords -> {
+            for (List<StreamRecordBatch> blockRecords : blocks) {
                 DataBlock block = new DataBlock(streamId, blockRecords);
                 waitingUploadBlocks.add(block);
                 waitingUploadBlocksSize += block.size();
-            });
+            }
             if (waitingUploadBlocksSize >= partSizeThreshold) {
                 tryUploadPart();
             }

--- a/s3stream/src/main/java/com/automq/stream/s3/ObjectWriter.java
+++ b/s3stream/src/main/java/com/automq/stream/s3/ObjectWriter.java
@@ -97,7 +97,7 @@ public interface ObjectWriter {
             writer = objectStorage.writer(ObjectStorage.WriteOptions.DEFAULT, objectKey);
         }
 
-        public void write(long streamId, List<StreamRecordBatch> records) {
+        public synchronized void write(long streamId, List<StreamRecordBatch> records) {
             List<List<StreamRecordBatch>> blocks = groupByBlock(records);
             blocks.forEach(blockRecords -> {
                 DataBlock block = new DataBlock(streamId, blockRecords);
@@ -127,7 +127,7 @@ public interface ObjectWriter {
             return blocks;
         }
 
-        private synchronized void tryUploadPart() {
+        private void tryUploadPart() {
             for (; ; ) {
                 List<DataBlock> uploadBlocks = new ArrayList<>(waitingUploadBlocks.size());
                 boolean partFull = false;

